### PR TITLE
Fix: #732: Added StoreReview to request a review from the user.

### LINKED ISSUE #732: Add App Review API in iOS
### Description: 
Add the standard Apple App Review API to our iOS app.

Please post the configurable variables here and we can decide on what works best for our needs

### PR DESCRIPTION
**Pull request by Brandon-T**
- Fixes: https://github.com/brave/brave-ios/issues/732
- Adds ability to request a review from the user via StoreKit.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Test Plan:

There's no way to QA it for now. We only show app review prompt in release builds.

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -3104,4 +3110,43 @@ extension BrowserViewController {
             NavigationPath.handle(nav: path, with: self)
         }
     }
+    
+    func requestReviewIfNecessary() {
+        #if DEBUG
```
Comment (kylehickinson) [inline Client/Frontend/Browser/BrowserViewController.swift]: Do we want these requests to show up in beta builds? Maybe we should use `!AppConstants.BuildChannel.isRelease` instead here?

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 2]
```diff
@@ -3104,4 +3110,43 @@ extension BrowserViewController {
...
+        return
+        #endif
+        
+        let launchCount = Preferences.Review.launchCount.value
+        let threshold = Preferences.Review.threshold.value
+        
+        if threshold == 0 {
+            Preferences.Review.threshold.value = 14
```
Comment (kylehickinson) [inline Client/Frontend/Browser/BrowserViewController.swift]: Is there a reason we don't just use 14 as the default threshold value when making the preference above?

Comment (Brandon-T) [reply Client/Frontend/Browser/BrowserViewController.swift]: Fixed.

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 1 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/BrowserViewController.swift]: Fixed.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 3]
```diff
@@ -2994,6 +3000,40 @@ extension BrowserViewController {
             NavigationPath.handle(nav: path, with: self)
         }
     }
+    
+    func requestReviewIfNecessary() {
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController.swift]: Can we move this logic outside of bvc?
I think `BraveShared/Analytics/AppReview.swift` would be the best.

And we could add some unit tests to it, add `date: Date = Date()` argument to the method to make testing easier.
See `DAU.swift` and `DAUTests.swift` it should be kind of similar logic.

The only thing we should do in bvc is to request the review so perhaps this method could also return a boolean

File: Client/Application/ClientPreferences.swift [hunk 4]
```diff
@@ -78,5 +78,13 @@ extension Preferences {
         /// The toggles states for clear private data screen
         static let clearPrivateDataToggles = Option<[Bool]>(key: "privacy.clear-data-toggles", default: [])
     }
+    final class Review {
```
Comment (iccub) [inline Client/Application/ClientPreferences.swift]: If we move `requestReviewIfNecessary` to `BraveShared`, this needs to be moved to `BraveShared/Preferences.swift`

File: Client/Application/ClientPreferences.swift [hunk 5]
```diff
@@ -78,5 +78,13 @@ extension Preferences {
...
+        /// Application Launch Count (how many times the application has been launched)
+        static let launchCount = Option<Int>(key: "review.launch-count", default: 0)
+        /// Review Threshold (the total amount of launches needed for the next review to show up)
+        static let threshold = Option<Int>(key: "review.threshold", default: 14)
```
Comment (iccub) [inline Client/Application/ClientPreferences.swift]: We should have a static constants for all three thresholds, put it in `AppReview.swift`

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 6]
```diff
@@ -2994,6 +3000,40 @@ extension BrowserViewController {
...
+        
+        let minimumDaysBetweenReviewRequest = 60
+        var daysSinceLastRequest = 0
+        if let previousRequest = Preferences.Review.lastReviewDate.value {
+            daysSinceLastRequest = Calendar.current.dateComponents([.day], from: previousRequest, to: Date()).day ?? 0
+        } else {
+            daysSinceLastRequest = minimumDaysBetweenReviewRequest
+        }
+        
+        if launchCount <= threshold ||  daysSinceLastRequest < minimumDaysBetweenReviewRequest {
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController.swift]: remove extra space `||  daysSinceLastRequest`

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 7]
```diff
@@ -427,6 +428,11 @@ class BrowserViewController: UIViewController {
         view.addInteraction(dropInteraction)
         
         initializeSyncWebView()
+        
+        //Request Review when the main-queue is free or on the next cycle.
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController.swift]: Add space after `//`

**Review by iccub — DISMISSED** [review]
Looks very nice, I would just move it to `BraveShared` and add a test or two

**Review by jhreis — APPROVED** [review]

**Pull request merged**